### PR TITLE
 feat: Show similar events by location 

### LINKED ISSUE #207: Filter events based on location in similar events section
To be done after #198 gets merged
Working on it :+1:

### PR DESCRIPTION
**Pull request by dreadpool2**
Fixes #207

### CODE REVIEW

File: app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsViewModel.kt [hunk 1]
```diff
@@ -2,27 +2,35 @@ package org.fossasia.openevent.general.event.topic
...
 import android.arch.lifecycle.ViewModel
+import android.text.TextUtils
 import io.reactivex.android.schedulers.AndroidSchedulers
 import io.reactivex.disposables.CompositeDisposable
 import io.reactivex.schedulers.Schedulers
+import org.fossasia.openevent.general.R.string.location
+import org.fossasia.openevent.general.data.Preference
 import org.fossasia.openevent.general.event.Event
 import org.fossasia.openevent.general.event.EventService
+import org.fossasia.openevent.general.utils.nullToEmpty
```
Comment (nikit19) [inline app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsViewModel.kt]: Extra import ?

File: app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsViewModel.kt [hunk 2]
```diff
@@ -2,27 +2,35 @@ package org.fossasia.openevent.general.event.topic
 
 import android.arch.lifecycle.MutableLiveData
 import android.arch.lifecycle.ViewModel
+import android.text.TextUtils
```
Comment (nikit19) [inline app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsViewModel.kt]: Extra import?

File: app/src/main/java/org/fossasia/openevent/general/event/EventService.kt [hunk 3]
```diff
@@ -63,20 +63,7 @@ class EventService(private val eventApi: EventApi, private val eventDao: EventDa
...
-                        .toFlowable()
-                        .map {
-                            eventDao.insertEvents(it)
-                        }
-                        .flatMap {
-                            eventsFlowable
-                        }
-        }
+    fun getSimilarEvents(id: Long, savedLocation: String): Single<List<Event>> {
+        return eventTopicApi.getEventsUnderTopicId(id, "name", savedLocation)
```
Comment (simarsingh24) [inline app/src/main/java/org/fossasia/openevent/general/event/EventService.kt]: why are you not inserting it in db?

File: app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsViewModel.kt [hunk 4]
```diff
@@ -2,27 +2,35 @@ package org.fossasia.openevent.general.event.topic
...
+    val savedLocation by lazy { preference.getString(tokenKey) }
     var eventId: Long = -1
 
     fun loadSimilarEvents(id: Long) {
         if(id == -1L){
             return
         }
-        compositeDisposable.add(eventService.getSimilarEvents(id)
+        val queryLocation = "[{\"name\":\"location-name\",\"op\":\"ilike\",\"val\":\"%$savedLocation%\"}]"
+
```
[the remark this answers is not in the document]
Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsViewModel.kt]: The feed is required I guess. We are giving a line feed space in Events View Model too

File: app/src/main/java/org/fossasia/openevent/general/event/EventService.kt [hunk 5]
```diff
@@ -63,20 +63,12 @@ class EventService(private val eventApi: EventApi, private val eventDao: EventDa
...
-                            eventDao.insertEvents(it)
-                        }
-                        .flatMap {
-                            eventsFlowable
-                        }
+    fun getSimilarEvents(id: Long, savedLocation: String): Single<List<Event>> {
+        return eventTopicApi.getEventsUnderTopicId(id, "name", savedLocation).flatMap { apiList ->
+            val eventIds = apiList.map { it.id }.toList()
+            eventDao.getFavoriteEventWithinIds(eventIds).flatMap { favIds ->
+                updateFavorites(apiList, favIds)
```
Comment (nikit19) [inline app/src/main/java/org/fossasia/openevent/general/event/EventService.kt]: don't blindly copy code. Take your time before pushing it

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/event/EventService.kt]: Actually the same process is being done and that's why it looks like it is being blindly copied :smile: . We are loading events based on event topic and updating favorites by mapping. Isn't mapping required?

**Comment by simarsingh24** [comment]
Look at the code! Why is favorite there!

**Comment by simarsingh24** [comment]
Did you run this locally it would never work

**Comment by simarsingh24** [comment]
Don’t push without testing your code for once

File: app/src/main/java/org/fossasia/openevent/general/event/EventService.kt (hunk 5 above)
Comment (simarsingh24) [reply app/src/main/java/org/fossasia/openevent/general/event/EventService.kt]: why are you returning favorite events from db

Comment (simarsingh24) [reply app/src/main/java/org/fossasia/openevent/general/event/EventService.kt]: you have to display similar events now only those events which are favorite in similar events will be displayed

**Comment by simarsingh24** [comment]
@dreadpool2  can you confirm is it working?

**Comment by dreadpool2** [comment]
@simarsingh24 I was just going to say that. The mapping is working. The events which are actually favorited are appearing to be favorited even after reload. The only issue is that the process in an infinite loop. The events are loaded constantly

File: app/src/main/java/org/fossasia/openevent/general/event/EventService.kt [hunk 6]
```diff
@@ -63,20 +63,12 @@ class EventService(private val eventApi: EventApi, private val eventDao: EventDa
...
+            }
```
Comment (dreadpool2) [inline app/src/main/java/org/fossasia/openevent/general/event/EventService.kt]: @iamareebjamal The process is in an infinite loop. The same process occurs correctly for loading of events by location in main screen, but here even after the mapping is done correctly, the loading still occurs.

**Comment by simarsingh24** [comment]
@dreadpool2 could you please finalize this PR its not nice to keep a PR open for a month 👍

**Comment by dreadpool2** [comment]
Doing that :+1:

**Pull request closed without merging**